### PR TITLE
Support contextual logging

### PR DESCRIPTION
**Pull request by bells17**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

I've updated the klog functions used within csi-lib-utils to contextual logging functions, following the guidelines below:
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

- Added contextual logging checks using logcheck. (You can run logcheck with `make test` or `make logcheck`)

**Does this PR introduce a user-facing change?**:

```release-note
Added support for contextual logging.
```

### CODE REVIEW

**Comment by bells17** [comment]
/wg structured-logging
/priority important-longterm
/kind cleanup
/cc @pohly

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#what-to-include-in-the-pull-request

File: pkg/controller/csi_handler.go [hunk 1]
```diff
@@ -186,7 +187,14 @@ func (h *csiHandler) ReconcileVA() error {
 
 		// If ListVolumes Attached Status is different, add to shared workQueue.
 		if attachedStatus != found {
-			klog.Warningf("VA %s for volume %s has attached status %v but actual state %v. Adding back to VA queue for forced reprocessing", va.Name, volumeHandle, attachedStatus, found)
+			logger.Error(
+				nil,
+				"VolumeAttachment for volume has attached status but actual state. Adding back to VolumeAttachment queue for forced reprocessing",
```
Comment (jsafrane) [inline pkg/controller/csi_handler.go:192]: This sounds odd without the values in-line in the log message.
Perhaps:

```suggestion
				"VolumeAttachment attached status and actual state do not match. Adding back to VolumeAttachment queue for forced reprocessing",
```

File: pkg/controller/csi_handler.go [hunk 2]
```diff
@@ -220,134 +228,138 @@ func (h *csiHandler) consumeForceSync(vaName string) bool {
...
+	logger.V(4).Info("CSIHandler: finished processing")
 }
 
-func (h *csiHandler) syncAttach(va *storage.VolumeAttachment) error {
+func (h *csiHandler) syncAttach(ctx context.Context, va *storage.VolumeAttachment) error {
+	logger := klog.FromContext(ctx)
 	if !h.consumeForceSync(va.Name) && va.Status.Attached {
 		// Volume is attached and no force sync, there is nothing to be done.
-		klog.V(4).Infof("%q is already attached", va.Name)
+		logger.V(4).Info("VolumeAttachment is already attached")
```
Comment (jsafrane) [inline pkg/controller/csi_handler.go:256]: Again, `va.Name` is not logged

File: pkg/controller/controller.go [hunk 3]
```diff
@@ -250,29 +252,30 @@ func (ctrl *CSIAttachController) processFinalizers(pv *v1.PersistentVolume) bool
...
+			logger.V(3).Info("PersistentVolume was deleted, ignoring")
 			return
 		}
-		klog.Errorf("Error getting PersistentVolume %q: %v", pvName, err)
+		logger.Error(err, "Error getting PersistentVolume")
 		ctrl.pvQueue.AddRateLimited(pvName)
 		return
 	}
-	ctrl.handler.SyncNewOrUpdatedPersistentVolume(pv)
+	ctrl.handler.SyncNewOrUpdatedPersistentVolume(ctx, pv)
```
Comment (jsafrane) [inline pkg/controller/controller.go:278]: IMO a new context with `pvName` should be passed.

File: pkg/controller/controller.go [hunk 4]
```diff
@@ -199,33 +200,34 @@ func (ctrl *CSIAttachController) pvUpdated(old, new interface{}) {
...
 		ctrl.vaQueue.AddRateLimited(vaName)
 		return
 	}
 	if va.Spec.Attacher != ctrl.attacherName {
-		klog.V(4).Infof("Skipping VolumeAttachment %s for attacher %s", va.Name, va.Spec.Attacher)
+		logger.V(4).Info("Skipping VolumeAttachment for attacher", "attacher", va.Spec.Attacher)
 		return
 	}
-	ctrl.handler.SyncNewOrUpdatedVolumeAttachment(va)
+	ctrl.handler.SyncNewOrUpdatedVolumeAttachment(ctx, va)
```
Comment (jsafrane) [inline pkg/controller/controller.go:230]: IMO a new context with `vaName` should be used. SyncNewOrUpdatedVolumeAttachment does not log the `vaName`, yet, this ctx does not have it (or I missed something).

**Comment by bells17** [comment]
/hold

https://kubernetes.slack.com/archives/C09QZFCE5/p1715087356356999

**Comment by bells17** [comment]
@jsafrane Thank you for your review.
I have made the changes. Could you please review it again?

**Comment by jsafrane** [comment]
/lgtm
/approve
/hold cancel
I will replace hack/verify-logcheck.sh with the version in release-tools in a subsequent PR, together with Kubernetes update.

**Pull request merged**